crates/viewer/re_ui/src/filter_widget.rs

```diff
@@ -0,0 +1,276 @@
+use egui::{Color32, NumExt as _, Widget as _};
+use itertools::Either;
+
+use crate::{list_item, UiExt as _};
+
+/// State for the filter widget when it is toggled on.
+#[derive(Debug, Clone)]
+struct InnerState {
+    /// The filter query string.
+    filter_query: String,
+
+    /// This ID is recreated every time the filter is toggled and tracks the current filtering
+    /// session.
+    ///
+    /// This can be useful for client code to store session-specific state (e.g., the state of tree
+    /// collapsed-ness).
+    session_id: egui::Id,
+}
+
+impl Default for InnerState {
+    fn default() -> Self {
+        let mut random_bytes = [0u8; 8];
+        getrandom::getrandom(&mut random_bytes).expect("Couldn't get random bytes");
+
+        Self {
+            filter_query: String::new(),
+
+            // create a new session id each time the filter is toggled
+            session_id: egui::Id::new(random_bytes),
+        }
+    }
+}
+
+/// State and UI for the filter widget.
+///
+/// The filter widget is designed as a toggle between a title widget and the filter text field.
+/// [`Self`] is responsible for storing the widget state as well as the query text typed by the
+/// user. [`FilterMatcher`] performs the actual filtering.
+#[derive(Debug, Clone, Default)]
+pub struct FilterState {
+    /// The current state of the filter widget.
+    ///
+    /// This is `None` when the filter is not active.
+    inner_state: Option<InnerState>,
+
+    /// Should the text field be focused?
+    ///
+    /// Set to `true` upon clicking on the search button.
+    request_focus: bool,
+}
+
+impl FilterState {
+    /// Return the filter if any.
+    ///
+    /// The widget must be enabled _and_ the filter string must not be empty.
+    pub fn query(&self) -> Option<&str> {
+        self.inner_state
+            .as_ref()
+            .map(|state| state.filter_query.as_str())
+    }
```

Comment:
comment and code aren't the same: if I have an empty query, I get still `Some` -> which then leads to a `Some("")` filter matcher which is distinct of of a `None` filter matcher which never gets hit.